pkg/templates/updateconfig/template.go

```diff
@@ -0,0 +1,156 @@
+package updateconfig
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"golang.stackrox.io/kube-linter/internal/stringutils"
+	"golang.stackrox.io/kube-linter/pkg/check"
+	"golang.stackrox.io/kube-linter/pkg/config"
+	"golang.stackrox.io/kube-linter/pkg/diagnostic"
+	"golang.stackrox.io/kube-linter/pkg/extract"
+	"golang.stackrox.io/kube-linter/pkg/lintcontext"
+	"golang.stackrox.io/kube-linter/pkg/objectkinds"
+	"golang.stackrox.io/kube-linter/pkg/templates"
+	"golang.stackrox.io/kube-linter/pkg/templates/updateconfig/internal/params"
+
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	templateKey = "update-configuration"
+)
+
+func compareIntOrString(max, min string, actual *intstr.IntOrString) bool {
+	if len(max) == 0 && len(min) == 0 {
+		return true
+	}
+	if actual == nil {
+		return false
+	}
+	actualVal, err := intstr.GetValueFromIntOrPercent(actual, 100, false)
+	if err != nil {
+		return false
+	}
+	actualIsPercent := strings.Contains(actual.String(), "%")
+	if len(max) > 0 {
+		maxIntOrString := intstr.Parse(max)
+		maxIsPercent := strings.Contains(maxIntOrString.String(), "%")
+		if actualIsPercent != maxIsPercent {
+			return false
+		}
+		maxVal, err := intstr.GetValueFromIntOrPercent(&maxIntOrString, 100, false)
+		if err != nil {
+			return false
+		}
+		if actualVal > maxVal {
+			return false
+		}
+	}
+	if len(min) > 0 {
+		minIntOrString := intstr.Parse(min)
+		minIsPercent := strings.Contains(minIntOrString.String(), "%")
+		if actualIsPercent != minIsPercent {
+			return false
+		}
+		minVal, err := intstr.GetValueFromIntOrPercent(&minIntOrString, 100, false)
+		if err != nil {
+			return false
+		}
+		if actualVal < minVal {
+			return false
+		}
+	}
+	return true
+}
+
+func conditional(firstCond bool, firstStr string, secondCond bool, secondStr, bothStr string) string {
+	msg := ""
+	if firstCond {
+		msg = firstStr
+	}
+	if firstCond && secondCond {
+		msg += bothStr
+	}
+	if secondCond {
+		msg += secondStr
+	}
+	return msg
+
+}
+
+func needsRollingUpdateDefinition(p params.Params) bool {
+	isRolling, _ := regexp.MatchString("Rolling", p.StrategyTypeRegex)
+	return isRolling && (len(p.MinPodsUnavailable) > 0 || len(p.MaxPodsUnavailable) > 0 ||
+		len(p.MinSurge) > 0 || len(p.MaxSurge) > 0)
+}
+
+func init() {
+	templates.Register(check.Template{
+		HumanName:   "Update configuration",
+		Key:         templateKey,
+		Description: "Flag configurations that do not meet the specified update configuration",
+		SupportedObjectKinds: config.ObjectKindsDesc{
+			ObjectKinds: []string{objectkinds.DeploymentLike},
+		},
+		Parameters:             params.ParamDescs,
+		ParseAndValidateParams: params.ParseAndValidate,
+		Instantiate: params.WrapInstantiateFunc(func(p params.Params) (check.Func, error) {
+			compiledRegex, err := regexp.Compile(p.StrategyTypeRegex)
+			if err != nil {
+				return nil, errors.Wrapf(err, "invalid regex %s", p.StrategyTypeRegex)
+			}
```

Comment:
It may make sense to parse the `IntOrString`s from the params here too, so we can return an error at check instantiation time if any are wrong? Right now if it's defined incorrectly, we will just silently not fail any checks, which doesn't seem ideal.

Reply:
Added checks and tests for configuration